src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionCodeFixProvider.Rewriter.cs

```diff
@@ -128,11 +130,40 @@ public override ExpressionSyntax VisitAssignmentExpression(AssignmentExpressionS
                 return node.Right;
             }
 
-            private ExpressionSyntax RewriteStatements(SyntaxList<StatementSyntax> statements)
+            private ExpressionSyntax RewriteStatements(SyntaxList<StatementSyntax> statements, ITypeSymbol variableSymbolTypeOpt, SemanticModel semanticModel)
             {
                 Debug.Assert(statements.Count == 1 || statements.Count == 2);
                 Debug.Assert(!statements[0].IsKind(SyntaxKind.BreakStatement));
-                return Visit(statements[0]);
+                var rewrittenExpression = Visit(statements[0]);
+
+                if (statements[0].Kind() == SyntaxKind.ExpressionStatement)
+                {
+                    var expressionStatement = (ExpressionStatementSyntax)statements[0];
+                    SyntaxNode type = null;
+
+                    // In cases of simple assignment (e.g. x = y), we see if the right-hand side can be cast to the type of the left-hand side.
+                    if (expressionStatement.Expression.Kind() == SyntaxKind.SimpleAssignmentExpression)
+                    {
+                        // Case: x = new Y();
+                        var objCreationNode = expressionStatement.Expression.ChildNodes().Where(s => s.Kind() == SyntaxKind.ObjectCreationExpression).FirstOrDefault();
+                        if (objCreationNode != default(SyntaxNode))
+                        {
+                            type = objCreationNode.ChildNodes().Where(s => s.Kind() == SyntaxKind.IdentifierName).FirstOrDefault();
+                        }
+                        // Case: x = y;
+                        else
+                        {
+                            type = expressionStatement.Expression.ChildNodes().Where(s => s.Kind() == SyntaxKind.IdentifierName).FirstOrDefault();
+                        }
```

Comment:
all these ChildNodes/FirstOrDefault aren't good.  We  shoudl be checking the exact shapes we're trying to support.